js/src/admin/components/UserListPage.tsx

```diff
@@ -0,0 +1,392 @@
+      className: 'UserListPage',
+      icon: 'fas fa-users',
+      title: app.translator.trans('core.admin.user_list.title'),
+      description: app.translator.trans('core.admin.user_list.description'),
+    };
+  }
+
+  /**
+   * Asynchronously fetch the next set of users to be rendered.
+   *
+   * Returns an array of Users, plus the raw API payload.
+   *
+   * Uses the `this.numPerPage` as the response limit.
+   *
+   * @param offset Offset to pass to API
+   */
+  async loadNextUserSet(offset: number = 0, pageNumber) {
+    // You shouldn't ever be negative
+    if (offset < 0) offset = 0;
+
+    app.store
+      .find('users', {
+        page: {
+          limit: this.numPerPage,
+          offset,
+        },
+      })
+      .then((apiData: UsersApiResponse) => {
+        // Next link won't be present if there's no more data
+        this.moreData = !!apiData.payload.links.next;
+
+        let data = apiData;
+
+        delete data.payload;
+
+        this.pageData = data;
+        this.pageNumber = pageNumber;
+        this.isLoadingPage = false;
+
+        m.redraw();
+      })
+      .catch((err) => {
+        console.error(err);
+        this.pageData = [];
+      });
+  }
+
+  /**
+   * Build an item list of columns to show for each user.
+   *
+   * Each column in the list should be an object with keys `name` and `content`.
+   *
+   * `name` is a string that will be used as the column name.
+   * `content` is a function with the User model passed as the first and only argument.
+   *
+   * See `UserListPage.tsx` for examples.
+   */
+  columns(): ItemList {
+    const columns = new ItemList();
+
+    columns.add(
+      'id',
+      {
+        name: app.translator.trans('core.admin.user_list.grid.default_columns.id'),
+        content: (user: User) => user.id(),
+      },
+      100
+    );
+
+    columns.add(
+      'username',
+      {
+        name: app.translator.trans('core.admin.user_list.grid.default_columns.username'),
+        content: (user: User) => user.username(),
+      },
+      90
+    );
+
+    columns.add(
+      'joinDate',
+      {
+        name: app.translator.trans('core.admin.user_list.grid.default_columns.join_time'),
+        content: (user: User) => (
+          <span class="UserList-joinDate" title={user.joinTime()}>
+            {dayjs(user.joinTime()).format('LLL')}
+          </span>
+        ),
+      },
+      80
+    );
+
+    columns.add(
+      'groupBadges',
+      {
+        name: app.translator.trans('core.admin.user_list.grid.default_columns.group_badges'),
+        content: (user: User) => {
+          const badges = user.badges().toArray();
+
+          if (badges.length) {
+            return <ul className="DiscussionHero-badges badges">{listItems(badges)}</ul>;
+          } else {
+            return app.translator.trans('core.admin.user_list.grid.default_columns.group_badges_none');
+          }
+        },
+      },
+      70
+    );
+
+    columns.add(
+      'emailAddress',
```

Comment:
Toggling the visibility of an email, then switching page, leaves the original email in view rather than redrawing this component. I'll look for a fix for this.

Reply:
Add a key based on user ID (https://mithril.js.org/keys.html), this will force a redraw when the key changes.